File: libbeardybraille.py
errors = {
        0: 'No errors',
        1: 'Invalid index',
        2: 'No more room in line',
        3: 'Index must be an int',
        }

bascii = " A1B'K2L@CIF/MSP\"E3H9O6R^DJG>NTQ,*5<-U8V.%[$+X!&;:4\\0Z7(_?W]#Y)="

class braillecell:

    def __init__(self, value):
        if type(value) == int:
            self.setfromint(value)
        elif type(value) == list:
            self.setfromlist(value)
        elif type(value) == str:
            if len(value) == 1:
                value = ord(value)
            else:
                raise ValueError
            if value < 0x2800:
                self.setfromascii(value)
            else:
                self.setfromunicode(value)
        else:
            raise TypeError

    def setfromint(self, value):
        if value >= 64:
            raise ValueError
        elif value < 0:
            raise ValueError
        else:
            self.brailleint = value

    def setfromlist(self, value):
        dotnums = []
        for i in range(1, 7):
            if i in value:
                dotnums.append(i)
        self.brailleint = sum(map(lambda x: 2**(x-1), dotnums))

    def setfromascii(self, value):
        if chr(value) in bascii:
            self.brailleint = bascii.index(chr(value))
        else:
            raise ValueError

    def setfromunicode(self, value):
        if (value < 0x2840) and (value >= 0x2800):
            self.brailleint = value - 0x2800
        else:
            raise ValueError

    def __str__(self):
        return chr(0x2800 + self.brailleint)

    def __repr__(self):
        return f'{self.__class__.__name__}({self.brailleint})'

    def ascii(self):
        return bascii[self.brailleint]

class brailleline:

    def __init__(self, value, width=40):
        if type(width) != int:
            raise TypeError
        if width <= 0:
            raise ValueError
        self.width = width
        if type(value) != str:
            raise TypeError
        if len(value) > width:
            raise ValueError
        self.content = []
        for cell in value:
            self.content.append(braillecell(cell))

    def validateindex(self, i):
        if type(i) != int:
            return 3
        if len(self.content) >= self.width:
            return 2
        if i < 0:
            return 1
        return 0

    def insert(self, i, value):
        if error := self.validateindex(i):
            return error
        if i > len(self.content):
            return 1
        self.content.insert(i, braillecell(value))
        return 0

    def delete(self, i):
        if error := self.validateindex(i):
            return error
        del self.content[i]
        return 0

    def __str__(self):
        rstr = ''
        for cell in self.content:
            rstr += str(cell)
        return rstr

    def ascii(self):
        rstr = ''
        for cell in self.content:
            rstr += cell.ascii()
        return rstr

    def __repr__(self):
        return f'a braile line of width {self.width} containing {len(self.content)} braille cells.'

File: libbeardybraille.js
'use scrict';

// BraileCell constructor

function BrailleCell(value) {

  this.value; // must be int n; 0 <= n < 64
  let bascii = " A1B'K2L@CIF/MSP\"E3H9O6R^DJG>NTQ,*5<-U8V.%[$+X!&;:4\\0Z7(_?W]#Y)=";

  setFromArray = function(a) {
    let sum = 0;
    for (let i = 1; i <= 6; i++) {
      if (a.includes(i)) {
        sum += 2 ** (i - 1);
      }
    }
    return sum;
  }

  setFromAscii = function(value) {
    if (bascii.includes(value)) {
      return bascii.indexOf(value);
    } else {
      throw "Invalid character";
    }
  }

  setFromUnicode = function(value) {
    let bunicode = "";
    for (let i = 0; i < 64; i++) {
      bunicode += String.fromCodePoint(0x2800 + i);
    }
    if (bunicode.includes(value)) {
      return bunicode.indexOf(value);
    } else {
      throw "Invalid character";
    }
  }

  if (typeof(value) === "number") {
    if (!(Number.isInteger(value))) {
      throw "Must be an int";
    }
    if (value < 0) {
      throw "Number too low";
    }
    if (value > 63) {
      throw "Number too high";
    }
    this.value = value;
  } else if (Array.isArray(value)) {
    this.value = setFromArray(value);
  } else if (typeof(value) === "string") {
    if (value.length === 1) {
      if (value.charCodeAt(0) < 128) {
        this.value = setFromAscii(value);
      } else {
        this.value = setFromUnicode(value);
      }
    } else {
      throw "String must contain only a single character";
    }
  } else {
    throw "Invalid input";
  }
};

// BrailleCell method definitions

BrailleCell.prototype.ascii = function() {
  return " A1B'K2L@CIF/MSP\"E3H9O6R^DJG>NTQ,*5<-U8V.%[$+X!&;:4\\0Z7(_?W]#Y)="[this.value];
}

BrailleCell.prototype.braille = function() {
  return String.fromCodePoint(0x2800 + this.value);
}


// BrailleLine constructor function

function BrailleLine(content, width) {
  this.content = []; // an array of BrailleCell objects

  if (typeof(width) === "number") {
    if (Number.isInteger(width)) {
      if (width < 1) {
        throw "width must be positive";
      } else {
        this.width = width
      }
    } else {
      throw "width must be an int";
    }
  } else {
    throw "width must be an int";
  }

  if (typeof(content) === "string") {
    if (content.length > this.width) {
      throw "content too long for width";
    }
  } else {
    throw "content must be string";
  }

  for (const cell of content) {
    this.content.push(new BrailleCell(cell));
  }

}

// BrailleLine method definitions.

BrailleLine.prototype.ascii = function() {
  let rstr = "";
  for (const cell of this.content) {
    rstr += cell.ascii();
  }
  return rstr;
}

BrailleLine.prototype.braille = function() {
  let rstr = "";
  for (const cell of this.content) {
    rstr += cell.braille();
  }
  return rstr;
}
